crates/pathfinder/src/storage/schema/revision_0018.rs

```diff
@@ -0,0 +1,202 @@
+    -- Keys are represented as base64 encoded strings separated by space
+    keys TEXT,
+    data BLOB,
+    FOREIGN KEY(block_number) REFERENCES canonical_blocks(number) ON DELETE CASCADE
+)",
+        [],
+    )
+    .context("Creating new table")?;
+
+    tx.execute(
+        r"-- Copy rowids to be sure that starknet_events_keys still references valid rows
+INSERT INTO starknet_events_new (rowid,block_number,idx,transaction_hash,from_address,keys,data)
+    SELECT rowid,block_number,idx,transaction_hash,from_address,keys,data FROM starknet_events",
+        [],
+    )
+    .context("Copying data")?;
+
+    // CREATE INDEX starknet_events_block_number ON starknet_events(block_number);
+    // CREATE INDEX starknet_events_from_address ON starknet_events(from_address);
+
+    tx.execute("DROP TABLE starknet_events", [])
+        .context("Dropping old table")?;
+
+    tx.execute(
+        "ALTER TABLE starknet_events_new RENAME TO starknet_events",
+        [],
+    )
+    .context("Renaming new table")?;
+
+    // Re-create triggers.
+    tx.execute(
+        r"CREATE TRIGGER starknet_events_ai
+    AFTER INSERT ON starknet_events
+    BEGIN
+        INSERT INTO starknet_events_keys(rowid, keys)
+        VALUES (
+            new.rowid,
+            new.keys
+        );
+    END;",
+        [],
+    )
+    .context("Creating after insert trigger")?;
+
+    tx.execute(
+        r"CREATE TRIGGER starknet_events_ad
+    AFTER DELETE ON starknet_events
+    BEGIN
+        INSERT INTO starknet_events_keys(starknet_events_keys, rowid, keys)
+        VALUES (
+            'delete',
+            old.rowid,
+            old.keys
+        );
+    END;",
+        [],
+    )
+    .context("Creating delete trigger")?;
+
+    tx.execute(
+        r"CREATE TRIGGER starknet_events_au
+    AFTER UPDATE ON starknet_events
+    BEGIN
+        INSERT INTO starknet_events_keys(starknet_events_keys, rowid, keys)
+        VALUES (
+            'delete',
+            old.rowid,
+            old.keys
+        );
+        INSERT INTO starknet_events_keys(rowid, keys)
+        VALUES (
+            new.rowid,
+            new.keys
+        );
+    END;",
+        [],
+    )
+    .context("Creating update trigger")?;
+
+    // Re-create indexs
+    tx.execute(
+        "CREATE INDEX starknet_events_block_number ON starknet_events(block_number)",
+        [],
+    )
+    .context("Creating block_number index")?;
+    tx.execute(
+        "CREATE INDEX starknet_events_from_address ON starknet_events(from_address)",
+        [],
+    )
+    .context("Creating from_address index")?;
+
+    Ok(())
+}
+
+/// Creates a new `canonical_blocks` table which holds the canonical StarkNet block chain.
+fn create_canonical_chain(tx: &Transaction<'_>) -> anyhow::Result<()> {
+    tx.execute(
+        r"-- Holds StarkNet's current canonical chain of blocks.
+CREATE TABLE canonical_blocks (
+    number INTEGER PRIMARY KEY NOT NULL,
+    hash   BLOB    NOT NULL,
```

Comment:
This could be unique but most likely it doesn't matter. This will need an index however if queries are made toward this but it can be changed later on of course, and this will be fast to rebuild.